Batteries/Data/Char/Basic.lean

```diff
@@ -22,3 +26,102 @@ theorem toNat_ofNat (n : Nat) : toNat (ofNat n) = if n.isValidChar then n else 0
   split
   · simp [ofNat, *]
   · simp [ofNat, toNat, *]
+
+/-- Returns `true` if `p` returns true for every `Char`. -/
+protected def all (p : Char → Bool) : Bool :=
+  -- 0x10FFFF (1114111) is the maximal valid code point.
+  Nat.all 0x110000 fun c h =>
+    -- Surrogate code points are from 0xD800 (55296) to 0xDFFF (57343).
+    if h₁ : c < 0xD800 then
+      p <| Char.ofNatAux c <| .inl h₁
+    else if h₂ : 0xDFFF < c then
+      p <| Char.ofNatAux c <| .inr ⟨h₂, h⟩
+    else
+      true
```

Comment:
This might be slightly faster:
```suggestion
  (Nat.all 0xD800 fun c h => p <| Char.ofNatAux c <| .inl h) &&
  (Nat.all (0x110000 - 0xE000) fun c h => p <| Char.ofNatAux (c + 0xE000) <| .inr <| by omega)
```
though a version of `Nat.all` with a lower bound would obviously be more readable.

Reply:
In what way is it faster?

Reply:
Of course! I will fix that.

Reply:
I would guess there's also some performence to win by iterating over UInt32s instead of Nat.

Reply:
Indeed, this is more like 0.36s:
```lean4
@[specialize]
def UInt32.allRange (a b : UInt32) (p : ∀ x, a ≤ x → x < b → Bool) : Bool :=
  if h : a < b then
    have : (b - (a + 1)).toNat < (b - a).toNat := sorry
    p a (Nat.le_refl _) h && UInt32.allRange (a + 1) b fun x ha hb => p x (sorry) hb
  else
    true
termination_by b - a

def all4 (p : Char → Bool) : Bool :=
  (UInt32.allRange 0 0xD800 fun c _ h => p <| Char.mk c <| .inl h) &&
  (UInt32.allRange 0xE000 0x110000 fun c h₁ h₂ => p <| Char.mk c <| .inr sorry)
```